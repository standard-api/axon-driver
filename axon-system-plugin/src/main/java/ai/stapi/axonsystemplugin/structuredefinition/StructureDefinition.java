package ai.stapi.axonsystemplugin.structuredefinition;

import ai.stapi.axonsystem.graphaggregate.AggregateWithDynamicGraph;
import ai.stapi.graphoperations.objectGraphMapper.model.MissingFieldResolvingStrategy;
import ai.stapi.graphsystem.dynamiccommandprocessor.DynamicCommandProcessor;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.FixedImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.StructureDefinitionImported;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.schema.structuredefinition.StructureDefinitionNormalizer;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateCreationPolicy;
import org.axonframework.modelling.command.CreationPolicy;
import org.axonframework.spring.stereotype.Aggregate;
import org.springframework.beans.factory.annotation.Autowired;

@Aggregate
public class StructureDefinition extends AggregateWithDynamicGraph<StructureDefinitionId> {

  @CommandHandler
  @CreationPolicy(AggregateCreationPolicy.CREATE_IF_MISSING)
  public void handle(
      ImportStructureDefinition command,
      @Autowired DynamicCommandProcessor commandProcessor
  ) {
    this.commandProcessor = commandProcessor;
    this.processCommandDynamically(
        new FixedImportStructureDefinition(
            command.getTargetIdentifier(),
            StructureDefinitionNormalizer.normalize(command.getStructureDefinitionSource())
        ),
        MissingFieldResolvingStrategy.LENIENT
    );
  }

  @EventSourcingHandler
  public void on(StructureDefinitionImported event) {
    this.onAggregateCreated(event);
  }
}