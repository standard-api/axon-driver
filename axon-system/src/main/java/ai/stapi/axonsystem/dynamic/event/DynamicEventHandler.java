package ai.stapi.axonsystem.dynamic.event;

import ai.stapi.axonsystem.dynamic.messagehandler.DynamicMessageHandler;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.axonframework.eventhandling.EventMessage;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.ANNOTATION_TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD})
@DynamicMessageHandler(
    messageType = EventMessage.class,
    payloadType = DynamicGraphUpdatedEvent.class
)
public @interface DynamicEventHandler {

  /**
   * Specifies the name of message that can be handled by the member method.
   */
  String messageName();

}
