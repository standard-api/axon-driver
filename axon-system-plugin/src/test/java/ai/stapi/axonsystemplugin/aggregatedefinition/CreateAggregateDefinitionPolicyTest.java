package ai.stapi.axonsystemplugin.aggregatedefinition;

import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import ai.stapi.graphsystem.structuredefinition.command.importStructureDefinitionFromSource.ImportStructureDefinition;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionData;
import ai.stapi.schema.structuredefinition.RawStructureDefinitionElementDefinition;
import ai.stapi.schema.structuredefinition.StructureDefinitionId;
import ai.stapi.test.domain.DomainTestCase;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.test.context.ActiveProfiles;

@ActiveProfiles("dev")
class CreateAggregateDefinitionPolicyTest extends DomainTestCase {

  @Test
  void itWillNotReactToComplexStructure() {
    this.whenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("ExampleComplex"),
            new RawStructureDefinitionData(
                "ExampleComplex",
                "http://some.url/ExampleComplex",
                "draft",
                "Example Description",
                "complex-type",
                false,
                "ExampleComplex",
                null,
                new RawStructureDefinitionData.Differential(
                    new ArrayList<>(List.of(
                        new RawStructureDefinitionElementDefinition(
                            "ExampleComplex.exampleField",
                            1,
                            "1",
                            "Example field description",
                            "Example field description",
                            "Example field description",
                            new ArrayList<>(List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            ))
                        )
                    ))
                )
            )
        )
    );
    var maybeEvent = this.getLastEventOfTypeOptionally(DynamicGraphUpdatedEvent.class);
    Assertions.assertTrue(maybeEvent.isEmpty());
  }

  @Test
  void itWillCreateAggregateDefinitionWhenResourceAdded() {
    this.whenCommandIsDispatched(
        new ImportStructureDefinition(
            new StructureDefinitionId("TestExampleResource"),
            new RawStructureDefinitionData(
                "TestExampleResource",
                "http://some.url/TestExampleResource",
                "draft",
                "Example Description",
                "resource",
                false,
                "TestExampleResource",
                null,
                new RawStructureDefinitionData.Differential(
                    new ArrayList<>(List.of(
                        new RawStructureDefinitionElementDefinition(
                            "TestExampleResource.exampleField",
                            1,
                            "1",
                            "Example field description",
                            "Example field description",
                            "Example field description",
                            new ArrayList<>(List.of(
                                new RawStructureDefinitionElementDefinition.ElementDefinitionType(
                                    "string",
                                    new ArrayList<>()
                                )
                            ))
                        )
                    ))
                )
            )
        )
    );

    this.thenLastDynamicEventOfNameApproved("AggregateDefinitionCreated");
  }
}