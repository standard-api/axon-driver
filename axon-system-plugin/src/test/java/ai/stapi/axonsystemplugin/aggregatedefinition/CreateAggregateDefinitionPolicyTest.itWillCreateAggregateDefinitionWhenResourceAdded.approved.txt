nodes:
	node_type: AggregateDefinition
	node_hash: 15F74
	node_edges:
		AggregateDefinition -> structure -> StructureDefinition (15F74 -> 45F30 -> 7380B)
	attributes:
		description -> Example Description
		name -> TestExampleResource

	node_type: StructureDefinition
	node_hash: 7380B
	node_edges:
		AggregateDefinition -> structure -> StructureDefinition (15F74 -> 45F30 -> 7380B)
	attributes:

edges:
	edge_type: structure
	edge_hash: 45F30
	AggregateDefinition -> structure -> StructureDefinition (15F74 -> 45F30 -> 7380B)
	attributes:
